Add method to easily mass-create regions

coasterteam opened the pull request:
A good way of pushing devs towards a likely easier way to setup a large amount of zones. Updated the README to include this as well. Follows code type-checking and format.

TheNexusAvenger commented on README.md:
````diff
@@ -67,6 +67,25 @@ local IsVisible = RegionState:IsRegionVisible("Region1") --Bool for if a region
 local InRegion = RegionState:IsInRegion("Region1", Vector3.new()) --Bool for if a point is in a region
 ```
 
+## Quicker Region Setup
````
I wouldn't have this as the same type of subsection as `RegionState` and `ModelCulling` when it should be part of `RegionState` or under it.

TheNexusAvenger commented on README.md:
```diff
@@ -67,6 +67,25 @@ local IsVisible = RegionState:IsRegionVisible("Region1") --Bool for if a region
+local RegionCulling = require(game:GetService("ReplicatedStorage"):WaitForChild("RegionCulling"))
+local RegionState = RegionCulling.RegionState
+
+RegionState:InsertRegionsFromInstance(game.ServerStorage.Regions)
```
Everything runs on the client. Referencing `ServerStorage` will be... problematic. Also should use `GetService` to be consistent with the other examples.

TheNexusAvenger commented on src/State/RegionState.lua:
```diff
@@ -95,6 +95,29 @@ function RegionState:AddRegion(RegionName: string, Center: CFrame, Size: Vector3
+		end
+	end
+	--// TODO
+	--// Perform region checks to maybe auto-connect intersecting regions? 
```
I don't really see the need to handle checking for intersections. You may actually want that like.

TheNexusAvenger commented on src/State/RegionState.lua:
```diff
@@ -95,6 +95,29 @@ function RegionState:AddRegion(RegionName: string, Center: CFrame, Size: Vector3
+		--Get the region name from the child name
+		local RegionName = RegionParent.Name
+		for _, RegionPart in RegionParent:GetChildren() do 
+			if RegionPart:IsA("BasePart") then
```
Not requesting a change or anything - just mentioning the rest of the code uses a lot of `if not [condition] then continue end`. It makes the code a bit less like a Christmas tree.

TheNexusAvenger reviewed (commented):
The code of the changes is fine, but I am not sure about the concept. Innovation Inc Thermal Power Plant stores all regions as data with the bounding boxes made from the relevant models. This design requires you to manage your regions along the models, which may be good or bad. You also then need to manage *another* source of truth to manage which regions can see which.

A unit test for this behavior would also be requested.

coasterteam commented:
> Innovation Inc Thermal Power Plant stores all regions as data with the bounding boxes made from the relevant models. This design requires you to manage your regions along the models, which may be good or bad.

That is a good way of handling setup. Basing loads on camera location along with zone intersections makes it easier to subdivide the regions. This could be converted to automatically add regions based on model bounding boxes instead of requiring you to build your own zones like this would add. Both could co-exist, but there's a cross-over of use cases.

TheNexusAvenger commented:
I am still not sure how to feel about dealing with 2 (or 3 if you include the models) sources of truth. You will have your region stored as parts and still need to store the connections somewhere else. The README doesn't provide any guidance on this.

coasterteam commented:
There's probably a better way to handle the For loop in https://github.com/IITPP-Roblox/Region-Culling/pull/1/commits/a299d47ee66187413cd8411400739036939d6761, not completely used for strict types yet. 

Added a new test case (it does pass!), updated types, tried to re-elaborate on README (still may need to come back to it), updated RegionState & BufferedRegionState to include the method. Created a working use case in my own world as well. 

Please let me know if there is still anything to resolve.

TheNexusAvenger commented on README.md:
```diff
@@ -39,8 +39,8 @@ RegionState:SetVisibleWhenOutsideRegions("Region3")
 --ConnectRegions makes it so both regions can see each other.
-RegionStateObject:ConnectRegions("Region1", "Region2")
-RegionStateObject:ConnectRegions("Region1", "Region3")
+RegionState:ConnectRegions("Region1", "Region2")
+RegionState:ConnectRegions("Region1", "Region3")
```
(No change) This was a good find. I completely missed that when copying from the unit test.

TheNexusAvenger commented on test/State/RegionState.spec.lua:
```diff
@@ -142,5 +142,19 @@ return function()
+            RegionStateObject:InsertRegionsFromInstance(InstanceTree)
+            expect(RegionStateObject:IsInRegion("Region4", Vector3.new(0, 1, 0))).to.equal(false)
+            expect(RegionStateObject:IsInRegion("Region4", Vector3.new(0, 3, 0))).to.equal(true)
+        end)
```
Would it be reasonable to add a test case for an empty instance? `expect(function() RegionStateObject:InsertRegionsFromInstance(Instance.new("Folder") end).to.throw("Instance \"Folder\" has no child instances, unable to build Regions from it.")` should work.

TheNexusAvenger reviewed (commented):
From the previous comment, I am still not sure about this idea. While it does regions, you will still need to manage a second source of truth for managing the connections.
